File: evals/harness_compare/lib/adapters/dry.py
from __future__ import annotations

import shutil
import time
from pathlib import Path

from ..task import Task
from .base import AdapterResult, BaseAdapter


class DryAdapter(BaseAdapter):
    """No agent: apply gold solution if present, else leave workspace as seeded.

    Use to validate graders. If `workspace_gold/` exists next to `workspace/`,
    copy it over the run workspace so check.py should PASS.
    """

    name = "dry"
    provides = {
        "terminal",
        "write_file",
        "read_file",
        "coding",
        "closed_form",
        "multi_step",
        "agent_os",
        "host_eval",
    }

    def run(self, task: Task, workspace: Path, instruction: str, timeout_s: int) -> AdapterResult:
        t0 = time.perf_counter()
        gold = task.path / "workspace_gold"
        if gold.is_dir():
            # replace workspace contents with gold
            for child in workspace.iterdir():
                if child.is_dir():
                    shutil.rmtree(child)
                else:
                    child.unlink()
            for item in gold.iterdir():
                dest = workspace / item.name
                if item.is_dir():
                    shutil.copytree(item, dest)
                else:
                    shutil.copy2(item, dest)
            detail = "applied workspace_gold"
        else:
            # closed_form: write expected answer file if tests/expected.txt exists
            expected = task.path / "tests" / "expected.txt"
            if expected.is_file():
                (workspace / "agent_answer.txt").write_text(
                    expected.read_text(encoding="utf-8"), encoding="utf-8"
                )
                detail = "wrote expected answer for closed_form"
            else:
                detail = "no gold; left seed workspace"
        # host_eval dry: write a success marker
        if "host_eval" in task.capabilities:
            (workspace / "host_eval_ok.txt").write_text("ok\n", encoding="utf-8")
            detail = "host_eval dry marker"
        t_ms = int((time.perf_counter() - t0) * 1000)
        _ = instruction, timeout_s
        return AdapterResult(status="ok", stdout=detail, t_ms=t_ms, detail=detail)

File: rust-agent/src/eval/compare/adapters/dry.rs
use super::{AdapterResult, HarnessAdapter};
use crate::eval::compare::task::{copy_dir_all, Task};
use std::fs;
use std::path::Path;
use std::time::Instant;

pub struct DryAdapter;

impl HarnessAdapter for DryAdapter {
    fn name(&self) -> &str {
        "dry"
    }

    fn provides(&self) -> &[&str] {
        &[
            "terminal",
            "write_file",
            "read_file",
            "coding",
            "closed_form",
            "multi_step",
            "agent_os",
            "host_eval",
        ]
    }

    fn run(&self, task: &Task, workspace: &Path, _instruction: &str) -> AdapterResult {
        let t0 = Instant::now();
        let gold = task.workspace_gold();
        let detail = if gold.is_dir() {
            // replace workspace with gold
            if let Ok(rd) = fs::read_dir(workspace) {
                for e in rd.flatten() {
                    let p = e.path();
                    if p.is_dir() {
                        let _ = fs::remove_dir_all(&p);
                    } else {
                        let _ = fs::remove_file(&p);
                    }
                }
            }
            if let Ok(rd) = fs::read_dir(&gold) {
                for e in rd.flatten() {
                    let src = e.path();
                    let dest = workspace.join(e.file_name());
                    if src.is_dir() {
                        let _ = copy_dir_all(&src, &dest);
                    } else {
                        let _ = fs::copy(&src, &dest);
                    }
                }
            }
            "applied workspace_gold".to_string()
        } else {
            let expected = task.path.join("tests").join("expected.txt");
            if expected.is_file() {
                if let Ok(text) = fs::read_to_string(&expected) {
                    let _ = fs::write(workspace.join("agent_answer.txt"), text);
                }
                "wrote expected answer for closed_form".into()
            } else {
                "no gold; left seed workspace".into()
            }
        };
        let detail = if task.capabilities.iter().any(|c| c == "host_eval") {
            let _ = fs::write(workspace.join("host_eval_ok.txt"), "ok\n");
            "host_eval dry marker".to_string()
        } else {
            detail
        };
        AdapterResult {
            status: "ok".into(),
            stdout: detail.clone(),
            stderr: String::new(),
            t_ms: t0.elapsed().as_millis() as u64,
            detail,
            capabilities_missing: vec![],
            exit_code: Some(0),
        }
    }
}
